proof/meta.go

```diff
@@ -85,6 +131,188 @@ func (m *MetaReveal) Validate() error {
+// display value. If the metadata is not JSON, or the decimal display key is
+// missing, the default decimal display value of 0 is returned.
+func (m *MetaReveal) GetDecDisplay() (map[string]interface{}, uint32, error) {
+	// An empty meta reveal has an implicit decimal display value of 0.
+	if m == nil {
+		return nil, 0, nil
+	}
+
+	if m.Type != MetaJson {
+		return nil, 0, ErrNotJSON
+	}
+
+	// Data must first be valid metadata JSON.
+	metaJSON, err := DecodeMetaJSON(m.Data)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Our default decimal display value is 0.
+	decDisplay, ok := metaJSON[MetadataDecDisplayKey]
+	if !ok {
+		return metaJSON, 0, ErrDecDisplayMissing
+	}
+
+	// We have to typecheck the value stored for the decimal display key;
+	// it could be another JSON type.
+	switch decDisplayVal := decDisplay.(type) {
+	case float64:
+		// Any Number-like values are stored in a float64. We need to
+		// range check the decoded value.
+		if decDisplayVal < 0 {
+			return nil, 0, fmt.Errorf("decoded decimal display "+
+				"value is negative: %v", decDisplayVal)
+		}
+
+		if decDisplayVal > float64(MaxDecDisplay) {
+			return nil, 0, fmt.Errorf("%w: %v",
+				ErrDecDisplayTooLarge, decDisplayVal)
+		}
+
+		return metaJSON, uint32(decDisplayVal), nil
+
+	default:
+		return nil, 0, ErrDecDisplayInvalidType
+	}
+}
+
+// SetDecDisplay attempts to set the decimal display value in existing JSON
+// metadata. It checks that the new metadata is below the maximum metadata size.
+func (m *MetaReveal) SetDecDisplay(decDisplay uint32) (*MetaReveal, error) {
+	err := IsValidDecDisplay(decDisplay)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch the current decimal display value.
+	currentMetaJSON, currentDecDisplay, err := m.GetDecDisplay()
+	switch {
+	// The current metadata is valid JSON. Either no decimal display value
+	// is present, or it is but doesn't match the desired value.
+	case err == ErrDecDisplayMissing, currentDecDisplay != decDisplay:
+		// If the requested decimal display value is 0, we don't need to
+		// add the field at all, as that is the default.
+		if decDisplay == 0 {
+			return &MetaReveal{
+				Type: m.Type,
+				Data: m.Data,
+			}, nil
+		}
+
+		// Otherwise, set the decimal display value and re-validate the
+		// JSON object.
+		updatedJSON := make(map[string]interface{})
+		maps.Copy(updatedJSON, currentMetaJSON)
+
+		updatedJSON[MetadataDecDisplayKey] = decDisplay
+
+		updatedJSONBytes, err := EncodeMetaJSON(updatedJSON)
+		if err != nil {
+			return nil, fmt.Errorf("invalid metadata after "+
+				"setting decimal display: %w", err)
+		}
+
+		return &MetaReveal{
+			Type: m.Type,
+			Data: updatedJSONBytes,
+		}, nil
+
+	// No metadata update needed.
+	case currentDecDisplay == decDisplay:
+		return &MetaReveal{
+			Type: m.Type,
+			Data: m.Data,
+		}, nil
+
+	// Our metadata is invalid in another way.
+	default:
+		return nil, fmt.Errorf("%w: %d", ErrDecDisplayTooLarge,
+			decDisplay)
```

Comment:
The error here seems very specific given the switch clause.

Reply:
+1, would expect a more generic `ErrInvalidDecDisplay` error ?